File: hostel_management_system_backend/Controllers/StudentPreferencesController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;

[ApiController]
[Route("api/student-preferences")]
[Authorize]
public sealed class StudentPreferencesController : ControllerBase
{
    private readonly IStudentPreferencesService _service;

    public StudentPreferencesController(IStudentPreferencesService service)
    {
        _service = service;
    }

    [HttpGet("me")]
    public async Task<ActionResult<StudentPreferenceReadDto>> GetMine(CancellationToken cancellationToken)
    {
        if (!TryGetCurrentUserId(out var userId))
            return Unauthorized("Invalid token.");

        var preference = await _service.GetMineAsync(userId, cancellationToken);
        return Ok(preference);
    }

    [HttpPut("me")]
    public async Task<ActionResult<StudentPreferenceReadDto>> UpsertMine([FromBody] StudentPreferenceUpsertDto dto, CancellationToken cancellationToken)
    {
        if (!TryGetCurrentUserId(out var userId))
            return Unauthorized("Invalid token.");

        var preference = await _service.UpsertMineAsync(userId, dto, cancellationToken);
        return Ok(preference);
    }

    private bool TryGetCurrentUserId(out Guid userId)
    {
        var sub = User.FindFirstValue(JwtRegisteredClaimNames.Sub)
            ?? User.FindFirstValue(ClaimTypes.NameIdentifier);

        return Guid.TryParse(sub, out userId);
    }
}

File: src/student-preferences/student-preferences.controller.ts
import { Body, Controller, Get, Put, UseGuards } from '@nestjs/common';

import { CurrentUserDecorator } from '../common/decorators/current-user.decorator';
import { JwtAuthGuard } from '../common/guards/jwt-auth.guard';
import { CurrentUser } from '../common/interfaces/current-user.interface';
import { StudentPreferenceReadDto, StudentPreferenceUpsertDto } from './dto/student-preferences.dto';
import { StudentPreferencesService } from './student-preferences.service';

@Controller('student-preferences')
@UseGuards(JwtAuthGuard)
export class StudentPreferencesController {
  constructor(private readonly studentPreferencesService: StudentPreferencesService) {}

  @Get('me')
  getMine(@CurrentUserDecorator() currentUser: CurrentUser): Promise<StudentPreferenceReadDto> {
    return this.studentPreferencesService.getMine(currentUser.userId);
  }

  @Put('me')
  upsertMine(
    @CurrentUserDecorator() currentUser: CurrentUser,
    @Body() dto: StudentPreferenceUpsertDto,
  ): Promise<StudentPreferenceReadDto> {
    return this.studentPreferencesService.upsertMine(currentUser.userId, dto);
  }
}
